core/src/controls.ts

```diff
@@ -16,82 +16,118 @@
+export type ControlSignalUserArgs =
+    BinaryCounterUserArgs | ChordProgressionUserArgs;
+
+/**
+ * Creates a `ControlSignal` based on the given `ControlSignalSpec`.
+ *
+ * @param type Specifies the `ControlSignal` to create.
+ * @returns A new `ControlSignal` object based on `spec`.
+ */
+export function controlSignalFromSpec(spec: BinaryCounterSpec): BinaryCounter;
+export function controlSignalFromSpec(spec: ChordProgressionSpec):
+    ChordProgression;
+export function controlSignalFromSpec(spec: ControlSignalSpec) {
+  switch (spec.type) {
+    case 'BinaryCounter':
+      return new BinaryCounter(spec.args);
+    case 'ChordProgression':
+      return new ChordProgression(spec.args);
+    default:
+      throw new Error(`Unknown control signal: ${spec}`);
+  }
+}
 
 /**
  * Abstract ControlSignal class for producing control tensors.
  *
- * @param numSteps The length of the control signal.
  * @param depth The size of the control tensor at each step.
  */
-export abstract class ControlSignal {
-  readonly numSteps: number;  // Total length of sequences.
-  readonly depth: number;     // Size of final output dimension.
-  abstract getTensor(step: number): tf.Tensor1D;
+export abstract class ControlSignal<T> {
+  readonly depth: number;  // Size of final output dimension.
+  abstract getTensors(numSteps: number, args: T): tf.Tensor2D;
 
-  constructor(numSteps: number, depth: number) {
-    this.numSteps = numSteps;
-    this.depth = depth;
-  }
+  constructor(depth: number) { this.depth = depth; }
 }
 
 /**
  * Binary counter control signal.
  *
- * @param numSteps The length of the control signal.
  * @param numBits The number of binary counter bits.
  */
-export class BinaryCounter extends ControlSignal {
-  constructor(numSteps: number, numBits: number) { super(numSteps, numBits); }
+export class BinaryCounter extends ControlSignal<BinaryCounterUserArgs> {
+  constructor(args: BinaryCounterArgs) { super(args.numBits); }
 
   /**
-   * Get the tensor of control values at the specified step. The returned tensor
-   * acts as a binary counter (using -1 instead of 0), with least significant
-   * bit first.
+   * Get the tensor of control values. The returned tensor acts as a binary
+   * counter (using -1 instead of 0), with least significant bit first.
    *
-   * @param step Step for which to get the binary counter.
+   * @param numSteps Number of steps of binary counter values to return.
+   * @returns A 2D tensor of binary counter values.
    */
-  getTensor(step: number) {
-    if (step >= this.numSteps) {
-      throw new RangeError(`Step out of range: ${step} >= ${this.numSteps}`);
+  getTensors(numSteps: number, args: BinaryCounterUserArgs) {
+    const buffer = tf.buffer([numSteps, this.depth]);
+    for (let step = 0; step < numSteps; ++step) {
+      for (let i = 0; i < this.depth; ++i) {
+        buffer.set(Math.floor(step / Math.pow(2, i)) % 2 ? 1.0 : -1.0, step, i);
+      }
```

Comment:
I think this may have an off-by-one issue, as [the binary counter is tracking the meter of the next step](https://github.com/tensorflow/magenta/blob/master/magenta/music/encoder_decoder.py#L522-L526).

Reply:
I guess it depends how this is supposed to align with the input sequence.  With chord conditioning, at each step the model should be fed melody[t-1] and chord[t], and output melody[t].  I thought it's the same for the binary counter...

Reply:
Yeah that sounds about right, similar to chords the binary counter is one step ahead of the input. If that's how this is set up then should be fine :)